controller/src/main/java/io/pravega/controller/store/stream/ZKStreamMetadataStore.java

```diff
@@ -60,4 +86,174 @@ ZKScope newScope(final String scopeName) {
+        Preconditions.checkNotNull(listener);
+
+        PathChildrenCacheListener bucketListener = (client, event) -> {
+            StreamImpl stream;
+            switch (event.getType()) {
+                case CHILD_ADDED:
+                    stream = getStreamFromPath(event.getData().getPath());
+                    listener.notify(new StreamNotification(stream.getScope(), stream.getStreamName(), NotificationType.StreamAdded));
+                    break;
+                case CHILD_REMOVED:
+                    stream = getStreamFromPath(event.getData().getPath());
+                    listener.notify(new StreamNotification(stream.getScope(), stream.getStreamName(), NotificationType.StreamRemoved));
+                    break;
+                case CHILD_UPDATED:
+                    stream = getStreamFromPath(event.getData().getPath());
+                    listener.notify(new StreamNotification(stream.getScope(), stream.getStreamName(), NotificationType.StreamUpdated));
+                    break;
+                case CONNECTION_LOST:
+                    listener.notify(new StreamNotification(null, null, NotificationType.ConnectivityError));
+                    break;
+                default:
+                    log.warn("Received unknown event {}", event.getType());
+            }
+        };
+
+        String bucketRoot = String.format(ZKStoreHelper.BUCKET_PATH, bucket);
+
+        bucketCacheMap.put(bucket, new PathChildrenCache(storeHelper.getClient(), bucketRoot, true));
+        PathChildrenCache pathChildrenCache = bucketCacheMap.get(bucket);
+        pathChildrenCache.getListenable().addListener(bucketListener);
+        pathChildrenCache.start(PathChildrenCache.StartMode.NORMAL);
+        log.info("bucket {} change notification listener registered", bucket);
+    }
+
+    @Override
+    public void unregisterBucketListener(int bucket) {
+        PathChildrenCache cache = bucketCacheMap.remove(bucket);
+        if (cache != null) {
+            try {
+                cache.clear();
+                cache.close();
+            } catch (IOException e) {
+                log.warn("unable to close watch on bucket {}", bucket);
+            }
+        }
+    }
+
+    @Override
+    public CompletableFuture<Boolean> takeBucketOwnership(int bucket, String processId, Executor executor) {
+        // try creating an ephemeral node
+        String bucketPath = ZKPaths.makePath(ZKStoreHelper.BUCKET_OWNERSHIP_PATH, String.valueOf(bucket));
+
+        return storeHelper.createEphemeralZNode(bucketPath, SerializationUtils.serialize(processId))
+                .thenCompose(created -> {
+                    if (!created) {
+                        // Note: data may disappear by the time we do a getData. Let exception be thrown from here
+                        // so that caller may retry.
+                        return storeHelper.getData(bucketPath)
```

Comment:
This is the case for "Data_Exists". 
createNode returns true if it has been created successfully, false if data exists. For any other exception, its thrown back. 
So if we get here, it means data exists, and we will try to retrieve and check if it has the same id as ours (in case this is a retry case).

Reply:
OK.